invoices/update.go

```diff
@@ -93,9 +102,14 @@ func updateInvoice(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
 			return nil, ctx.acceptRes(resultReplayToAccepted), nil
 
 		case channeldb.HtlcStateSettled:
+			keysendPreimage := ctx.keysendPreimage()
```

Comment:
Why's this method needed? Can't we just set `PaymentPreimage` properly in this context? Or is the idea here to slowly phase out that field?

Reply:
in my head, the `PaymentPreimage` won't be set for an AMP payment. there isn't even a logical value we could put there at the time the invoice is generated. so, we have to get the preimage from the invoice context and not the raw invoice. down the road there will probably be a more generic `preimage()` method that automatically returns the appropriate preimage based on the context (legacy or AMP), but for now we don't have that distinction.

Reply:
[added a godoc](https://github.com/lightningnetwork/lnd/pull/4338/files#diff-0f5a3cd97f21d51209c1d11db90f4e7dR48) to explain the rationale behind this method in particular